GH-27: Add allowSchemaUnionization config property

C0urante opened the pull request:
Addresses #27 by adding an `allowSchemaUnionization` config property that defaults to `false` and dictates whether schema unionization should occur during schema updates. Unit tests for the `SchemaManager` class (where the bulk of this logic is located) are added to verify expected behavior.

criccomini commented:
IIRC, we had some discussion on this. Do we actually need an explicit unionization param? My recollection was that if both of these are false `allowNewBQFields, allowReqFieldRelaxation` that it's equivalent to disallowing unionization.

criccomini commented:
Ah, I see your comment on https://github.com/confluentinc/kafka-connect-bigquery/issues/27 now. I think that explains the nuanced difference where both new fields and field relaxation are allowed, but you don't want unionization.

mtagle commented:
How are we determining if a new schema is totally disparate from an old schema or simply contains both newly nulled fields and new fields? or am I misunderstanding something?

C0urante commented:
@mtagle if schema unionization is disabled, then we validate schema changes by comparing the latest schema from Kafka for the table to the current schema for the table, and then checking to make sure that fields are relaxed and added as permitted by the connector's `allowNewBQFields` and `allowBQRequiredFieldRelaxation` config properties. Then, assuming those changes are permitted by the connector config, we try to update the BigQuery table with the latest incoming schema from Kafka. The key difference here is that we never re-add fields from the existing schema to the new schema, which has the effect that BigQuery does the validation for us to ensure that the two schemas aren't completely disjoint (since if they are, even if the existing schema has all `NULLABLE` fields, the new schema will be rejected since it will be completely missing those fields).

C0urante commented:
Thanks @mtagle and @stoynov96 :)

I've added unit tests now and, in the process, identified and fixed some bugs in the schema update logic. Would you mind giving this one more pass? I especially want to make sure I'm getting the unionization logic correct; I've outlined the expected behavior in some of the new unit test cases, and would appreciate a second look to make sure that this aligns with your expectations for that feature as well.

C0urante commented:
@mtagle @stoynov96 I've rebased onto the latest 2.0.x; I'll merge now but please feel free to file an issue if there's any unexpected change in schema unionization behavior. I'll be happy to follow up myself :)
